### PR TITLE
APEXMALHAR-2335 APEXMALHAR-2333 APEXMALHAR-2334 #resolve #comment Pro…

### PR DESCRIPTION
**Pull request by brightchen**
…blems on StateTracker

### CODE REVIEW

**Comment by brightchen** [comment]
@davidyan74 Please review

**Comment by brightchen** [comment]
@davidyan74 please review

File: library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java [hunk 1]
```diff
@@ -435,17 +435,46 @@ public void put(Slice key, long timeBucket, Slice value)
...
     @Override
     public long freeMemory(long windowId) throws IOException
     {
-      // calculate the size first and then send the release memory request. It could reduce the chance of conflict and increase the performance.
-      long size = keyStream.dataSizeUpToWindow(windowId) + valueStream.dataSizeUpToWindow(windowId);
+
+      long memoryFreed = 0;
+      Long clearWindowId;
+
+      while ((clearWindowId = committedData.floorKey(windowId)) != null) {
```
Comment (davidyan74) [inline library/src/main/java/org/apache/apex/malhar/lib/state/managed/Bucket.java]: each floorKey and remove call is O(log n) so the loop is going to be O(n log n). Why don't we just have a normal ordered iteration to make it O(n) instead?

File: library/src/main/java/org/apache/apex/malhar/lib/state/managed/StateTracker.java [hunk 2]
```diff
@@ -32,56 +35,55 @@
...
 
   protected transient AbstractManagedStateImpl managedStateImpl;
 
+  private transient long lastUpdateAccessTime = 0;
+  private transient Set<Long> accessedBucketIds = Sets.newHashSet();
+  private transient TreeSet<BucketIdTimeWrapper> bucketHeapForAccess = Sets.newTreeSet(new BucketIdTimeWrapper.BucketTimeComparator());
+  private transient TreeSet<BucketIdTimeWrapper> bucketHeapForRelease = Sets.newTreeSet(new BucketIdTimeWrapper.BucketTimeComparator());
+  private transient ReadWriteLock bucketAcessLock = new ReentrantReadWriteLock();
+  private int updateAccessTimeInterval = 500;
+
```
Comment (davidyan74) [inline library/src/main/java/org/apache/apex/malhar/lib/state/managed/StateTracker.java:54]: Can we make some of these members final?

File: library/src/main/java/org/apache/apex/malhar/lib/state/managed/StateTracker.java [hunk 3]
```diff
@@ -148,6 +179,19 @@ void teardown()
    */
   private static class BucketIdTimeWrapper
   {
+    public static class BucketTimeComparator implements Comparator<BucketIdTimeWrapper>
+    {
+      @Override
+      public int compare(BucketIdTimeWrapper o1, BucketIdTimeWrapper o2)
+      {
+        //same bucket are equals.
+        if (o1.bucketId == o2.bucketId) {
```
Comment (davidyan74) [inline library/src/main/java/org/apache/apex/malhar/lib/state/managed/StateTracker.java]: why do we need this if statement?

**Review by davidyan74 — CHANGES_REQUESTED** [review]

File: library/src/main/java/org/apache/apex/malhar/lib/state/managed/StateTracker.java [hunk 4]
```diff
@@ -32,56 +35,55 @@
...
-
   private final transient Timer memoryFreeService = new Timer();
 
   protected transient AbstractManagedStateImpl managedStateImpl;
 
+  private transient long lastUpdateAccessTime = 0;
+  private transient Set<Long> accessedBucketIds = Sets.newHashSet();
+  private transient TreeSet<BucketIdTimeWrapper> bucketHeapForAccess = Sets.newTreeSet(new BucketIdTimeWrapper.BucketTimeComparator());
+  private transient TreeSet<BucketIdTimeWrapper> bucketHeapForRelease = Sets.newTreeSet(new BucketIdTimeWrapper.BucketTimeComparator());
+  private transient ReadWriteLock bucketAcessLock = new ReentrantReadWriteLock();
```
Comment (davidyan74) [inline library/src/main/java/org/apache/apex/malhar/lib/state/managed/StateTracker.java]: spelling

File: library/src/main/java/org/apache/apex/malhar/lib/state/managed/StateTracker.java [hunk 5]
```diff
@@ -105,12 +107,30 @@ public void run()
...
+          for (BucketIdTimeWrapper releaseItem : bucketHeapForRelease) {
+            bucketHeapForAccess.add(releaseItem);
+          }
+          bucketHeapForRelease.clear();
+
+          //switch the access and release list
+          TreeSet<BucketIdTimeWrapper> tmp = bucketHeapForAccess;
+          bucketHeapForAccess = bucketHeapForRelease;
+          bucketHeapForRelease = tmp;
+        } finally {
```
Comment (davidyan74) [inline library/src/main/java/org/apache/apex/malhar/lib/state/managed/StateTracker.java]: Not sure what this code is doing. Is it the same as putting all items in bucketHeapForAccess to bucketHeapForRelease?

File: library/src/main/java/org/apache/apex/malhar/lib/state/managed/StateTracker.java [hunk 6]
```diff
@@ -32,56 +32,53 @@
...
   }
 
   void bucketAccessed(long bucketId)
   {
-    BucketIdTimeWrapper idTimeWrapper = bucketAccessTimes.get(bucketId);
-    if (idTimeWrapper != null) {
-      bucketHeap.remove(idTimeWrapper);
-    }  else {
-      idTimeWrapper = new BucketIdTimeWrapper(bucketId);
+    accessedBucketIds.add(bucketId);
```
Comment (davidyan74) [inline library/src/main/java/org/apache/apex/malhar/lib/state/managed/StateTracker.java]: accessBucketIds is not a ConcurrentSet. Won't this create a problem when multiple threads call this method?

Comment (brightchen) [reply library/src/main/java/org/apache/apex/malhar/lib/state/managed/StateTracker.java]: accessedBucketIds all used by operator main thread

File: library/src/main/java/org/apache/apex/malhar/lib/state/managed/StateTracker.java [hunk 7]
```diff
@@ -32,56 +32,53 @@
...
-
   private final transient Timer memoryFreeService = new Timer();
 
   protected transient AbstractManagedStateImpl managedStateImpl;
 
+  private transient long lastUpdateAccessTime = 0;
+  private final transient Set<Long> accessedBucketIds = Sets.newHashSet();
+  private transient TreeSet<BucketIdTimeWrapper> bucketHeapForAccess = Sets.newTreeSet();
+  private transient TreeSet<BucketIdTimeWrapper> bucketHeapForRelease = Sets.newTreeSet();
+
```
Comment (davidyan74) [inline library/src/main/java/org/apache/apex/malhar/lib/state/managed/StateTracker.java:52]: I think these two TreeSets and the BucketIdTimeWrapper can be replaced by a LinkedHashMap with accessOrder = true

Comment (brightchen) [reply library/src/main/java/org/apache/apex/malhar/lib/state/managed/StateTracker.java:52]: it is possible to it by a map from time to a list of bucket. but seems not simplify the code. The reason to use two lists in order two threads can run without concurrent control. the concurrent control only need to happen when merge/swap

Comment (davidyan74) [reply library/src/main/java/org/apache/apex/malhar/lib/state/managed/StateTracker.java:52]: Please take a look at this and let me know what you think:
https://github.com/apache/apex-malhar/commit/eadf39b11f495e84c1d1e6ffbb83c20b8dee404b

Comment (brightchen) [reply library/src/main/java/org/apache/apex/malhar/lib/state/managed/StateTracker.java:52]: it is possible that last update time is different. for example: in first 500ms, updated b1, b2; in second 500ms, updated b2, b3, so the last update of b1 is different from last update of b2 and b3

Comment (davidyan74) [reply library/src/main/java/org/apache/apex/malhar/lib/state/managed/StateTracker.java:52]: This is my latest FYI: https://github.com/davidyan74/apex-malhar/tree/bucketLastAccess

**Comment by brightchen** [comment]
@davidyan74 Merged your update, please review

**Pull request merged**